Optimize the performance of loading TouchStone files.

Asachoo opened the pull request:
Cache the `numbers_per_line` object and perform preliminary checks on strings to avoid unnecessary function executions.

Files of different sizes show a performance improvement of about **35%**.

The test results log is as follows:
```bash
(skrf) (base) ➜  scikit-rf git:(master) ✗ python load_performance.py
/home/xxx/scikit-rf/skrf/__init__.py
Time taken to load network 0.04 MB: 0.7483 ms (std: 0.0958 ms, min: 0.6616 ms)
Time taken to load network 1.37 MB: 43.2696 ms (std: 0.9668 ms, min: 41.9932 ms)
Time taken to load network 11.32 MB: 217.6757 ms (std: 15.4226 ms, min: 209.6951 ms)
Time taken to load network 78.85 MB: 1218.1206 ms (std: 42.5231 ms, min: 1183.3950 ms)
Time taken to load network 199.82 MB: 3229.8712 ms (std: 184.5385 ms, min: 3074.8850 ms)

(skrf) (base) ➜  scikit-rf git:(master) ✗ git switch TouchStoneFormat

(skrf) (base) ➜  scikit-rf git:(TouchStoneFormat) ✗ python load_performance.py 
/home/xxx/scikit-rf/skrf/__init__.py
Time taken to load network 0.04 MB: 0.5461 ms (std: 0.0830 ms, min: 0.4775 ms)
Time taken to load network 1.37 MB: 24.4713 ms (std: 3.3573 ms, min: 22.8682 ms)
Time taken to load network 11.32 MB: 141.6678 ms (std: 12.5191 ms, min: 134.8704 ms)
Time taken to load network 78.85 MB: 811.5176 ms (std: 45.9418 ms, min: 777.0676 ms)
Time taken to load network 199.82 MB: 2175.9645 ms (std: 164.8075 ms, min: 2027.6028 ms)
```

The test code is as follows:
```python3
import skrf as rf
import os
import numpy as np
from time import time

print(rf.__file__)

files = (
    "XXX.s2p", ...
)
ntimes = 20
for f in files:
    t = []
    for _ in range(ntimes):
        a = time() * 1000
        ntwk = rf.Network(f)
        b = time() * 1000
        t.append(b - a)
    t = np.array(t)
    print(
        f"Time taken to load network {os.path.getsize(f) / 1024**2:.2f} MB: {t.mean():.4f} ms (std: {t.std():.4f} ms, min: {t.min():.4f} ms)"
    )

```

FranzForstmayr commented on skrf/io/touchstone.py:
```diff
@@ -91,6 +92,19 @@ def n_ansys_impedance_values(self) -> int:
+        self._rank = x
+
+        # If the rank changes, 'numbers_per_line' needs to be recalculated.
+        if "numbers_per_line" in self.__dict__:
```
I'm not sure if this is even needed, the rank should definitely not change while parsing

Asachoo replied:
Thank you for your professional suggestion.
The `rank` property was mainly set to provide an entry point for **removing the `numbers_per_line` cache**. However, since you mentioned that the **rank should definitely not change**, the need to remove the cache becomes meaningless.
I will remove this part of the modification.

FranzForstmayr commented:
Maybe just this change would be enough:
```python
            line_l = line.lower()

            if line[0] in ["!", "#", "["]:
                for k, v in self._parse_dict.items():
                    if line_l.startswith(k):
                        v(line)
                        break
            else:
                ...
```

Asachoo commented:
## Summary
Thank you, @FranzForstmayr, for your insightful suggestions. I have summarized the changes made in this PR, which includes three modifications:
1. Caching the `numbers_per_line` variable using `cached_property`;
2. Pre-checking the character type of `line_l[0]`;
3. Using `list(map(float, ...))` to replace `list comprehensions`.

## Data
Below are the performance changes for each modification applied individually:

### 1. Master Branch (BaseLine)
```bash
Time taken to load network 0.04 MB: 0.7411 ms (std: 0.0923 ms, min: 0.6660 ms)
Time taken to load network 1.37 MB: 43.9980 ms (std: 2.8768 ms, min: 42.1416 ms)
Time taken to load network 11.32 MB: 216.2328 ms (std: 11.7919 ms, min: 206.6809 ms)
Time taken to load network 78.85 MB: 1230.7935 ms (std: 47.2551 ms, min: 1182.1987 ms)
Time taken to load network 199.82 MB: 3231.9390 ms (std: 185.4683 ms, min: 3038.6736 ms)
```

### 2. Set `numbers_per_line` to be `cached_property`
```bash
Time taken to load network 0.04 MB: 0.7005 ms (std: 0.0896 ms, min: 0.6438 ms)
Time taken to load network 1.37 MB: 37.1880 ms (std: 3.5364 ms, min: 35.4734 ms)
Time taken to load network 11.32 MB: 196.8984 ms (std: 14.4792 ms, min: 181.8379 ms)
Time taken to load network 78.85 MB: 1050.9514 ms (std: 47.1150 ms, min: 1015.2888 ms)
Time taken to load network 199.82 MB: 2746.3581 ms (std: 168.1199 ms, min: 2619.1348 ms)
```

### 3. Add a preliminary checks on strings, to keep the for..else structure valid:
```python3
...
continue_go = True
if line[0] in ["!", "#", "["]:
    for k, v in self._parse_dict.items():
        if line_l.startswith(k):
            v(line)
            continue_go = False
            break
if continue_go:
    values = [float(v) for v in line.partition("!")[0].split()]
    ...
```
Output:
```bash
Time taken to load network 0.04 MB: 0.6521 ms (std: 0.0881 ms, min: 0.5911 ms)
Time taken to load network 1.37 MB: 35.4366 ms (std: 3.9613 ms, min: 33.2773 ms)
Time taken to load network 11.32 MB: 177.2632 ms (std: 14.0723 ms, min: 169.9531 ms)
Time taken to load network 78.85 MB: 1033.8643 ms (std: 57.5237 ms, min: 997.0122 ms)
Time taken to load network 199.82 MB: 2739.5001 ms (std: 177.7936 ms, min: 2595.1284 ms)
```

### 4. Use `map(float, ...)` to replace list comprehensions `[float(v) for v in ...]`
```python3
...
if "!" in line:
    line = line.partition("!")[0]
values = list(map(float, line.split()))
...
```
Output:
```bash
Time taken to load network 0.04 MB: 0.8061 ms (std: 0.1930 ms, min: 0.6724 ms)
Time taken to load network 1.37 MB: 42.8646 ms (std: 3.4950 ms, min: 40.6724 ms)
Time taken to load network 11.32 MB: 211.8912 ms (std: 13.7295 ms, min: 205.5249 ms)
Time taken to load network 78.85 MB: 1157.2963 ms (std: 46.0382 ms, min: 1125.9507 ms)
Time taken to load network 199.82 MB: 3076.4666 ms (std: 168.6508 ms, min: 2908.4124 ms)
```
## Performance Analysis
The contribution of each of the three modifications to the performance change can be estimated:
1. Cached Property: (3231.93 -> 2746.35, delta: 485.58) => 485.58 / (485.58+492.43+155.47) = **43%** 
2. Preliminary check: (3231.93 -> 2739.50, delta: 492.43) => 492.43 / (485.58+492.43+155.47) = **43%**
3. map() function: (3231.93 -> 3076.46, delta: 155.47) => 155.47 / (485.58+492.43+155.47) = **14%**
4. 
## Conclusion
Each of these three modifications provided effective performance improvements.
Together, they contributed to the previously mentioned 35% reduction in load time.

FranzForstmayr commented:
Great improvement, thank you! :)
